Drivetrain clean

wollip opened the pull request:
Integrated changes from hardware overhaul

Commit:
Add changes from branch hardware-overhaul

Move old files

Commit:
Integrate hardware-overhaul changes

Update stages to match Drivetrain and Encoder interface

Change functionality of drivetrain

Implement beacon following in drivetrain

Update test code

Tune drivetrain

Remove LCD prints

Acceleration limit to motors

wollip commented on include/configs.h:
```diff
@@ -33,10 +33,9 @@
-#define GAIN_P() 12
-#define GAIN_I() 0
-#define GAIN_D() 0
+#define GAIN_P() 12.0
```
what are these use for?

wollip replied:
they are still being used by drivetrain

wollip commented on include/configs.h:
```diff
@@ -110,6 +110,21 @@
+#define LINE_FOLLOW_2_SPEED() 255
+#define LINE_FOLLOW_2_GAIN() 50
+
+/*#define LINE_FOLLOW_0_SPEED() 80
```
leave a message on why these are here

wollip commented on include/hardware/encoder.h:
```diff
@@ -1,43 +1,30 @@
 #include "configs.h"
-
 #include <stdint.h>
+#include "Print.h"
```
warp around with debug

wollip commented on include/hardware/encoder.h:
```diff
@@ -1,43 +1,30 @@
+  int32_t getPosition() const;
+  int16_t getVelocity() const;
+  void tick();
+  void printTo(Print& p);
```
debug

wollip commented on include/hardware/encodermotor.h:
```diff
@@ -0,0 +1,39 @@
+
+  enum ControlMode { IDLE, POWER, VELOCITY, POSITION };
+  ControlMode m_controlMode;
+  static const int16_t m_accelerationLimit = 20;
```
move to CPP

wollip commented on include/hardware/encodermotor.h:
```diff
@@ -0,0 +1,39 @@
+  int m_motor_num;
+  Encoder& m_encoder;
+
+  const static int32_t deadZone = 5;
```
move to CPP

wollip commented on include/sequences/drivetrain.h:
```diff
@@ -0,0 +1,53 @@
+  void setPower(int16_t left, int16_t right);
+  void setVelocity(int16_t left, int16_t right);
+  void tick();
+  void printTo(Print& p);
```
debug

wollip commented on include/utils/pid.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include <Print.h>
```
debug

wollip commented on include/utils/pid.h:
```diff
@@ -0,0 +1,22 @@
+  float kD;
+  float kDecay;
+  float run(float error);
+  void printTo(Print& p);
```
debug

wollip commented on src/hardware/encoder.cpp:
```diff
@@ -1,86 +1,63 @@
-  start(R_ENCODER_);
-  start(L_ENCODER_);
+int32_t Encoder::cmToTicks(float cm) {
+  return cm * GEAR_RATIO() * ENCODER_TICKS_PER_REVOLUTION() / (WHEEL_DIAMETER() * M_PI);
```
lets have this as a constant, so we don't have to do the calculation again and again.

wollip commented on src/hardware/encoder.cpp:
```diff
@@ -1,86 +1,63 @@
-  EICRA |= mode << (INTX*2);
-  sei();
+int32_t Encoder::degToTicks(float deg) {
+  return cmToTicks((deg / 180.) * AXLE_LENGTH() * M_PI);
```
change to constant

wollip commented on src/hardware/encoder.cpp:
```diff
@@ -1,86 +1,63 @@
 
-uint32_t Encoder::cmToTicks(unsigned int distance) {
-  return distance * GEAR_RATIO() * 24 / (WHEEL_DIAMETER() * PI);
+void Encoder::printTo(Print& p) {
```
debug

wollip commented on src/hardware/encoder.cpp:
```diff
@@ -1,86 +1,63 @@
-    hardware::counts[0]++;
-    hardware::prevTime[0] = millis();
+int32_t Encoder::getPositionPrivate() {
+  if (m_reverse) {
```
lets do this in 2 line?

wollip commented on src/hardware/encoder.cpp:
```diff
@@ -1,86 +1,63 @@
-    hardware::prevTime[3] = millis();
-  }
-};
+}
```
comment for namespace

wollip commented on src/hardware/encodermotor.cpp:
```diff
@@ -0,0 +1,85 @@
+#include "phys253.h"
+#include "configs.h"
+
+using namespace utils;
```
i don't like this

wollip commented on src/hardware/encodermotor.cpp:
```diff
@@ -0,0 +1,85 @@
+    }
+    case VELOCITY:
+      m_powerTarget += m_velocityPid.run(m_encoder.getVelocity() - m_velocityTarget);
+      if (m_powerTarget > m_maxPower) m_powerTarget = m_maxPower;
```
lets us `constrain` instead.

wollip commented on src/hardware/encodermotor.cpp:
```diff
@@ -0,0 +1,85 @@
+  setPosition(m_encoder.getPosition(), 20);
+}
+
+#define median(a,b,c) max(min(a,b), min(max(a,b),c))
```
take this out and replace with `constrain`

lsav commented on src/stages/gate.cpp:
```diff
@@ -43,14 +43,14 @@ bool Gate::loop(){
         }
 #endif  // CAUTIOUS_GATE_ROUTINE()
-        follower_.loop();
+        driver_.commandLineFollow(true);
```
commandLineFollow should not take a boolean?

wollip replied:
change `true` to `1`

wollip commented on src/hardware/encodermotor.cpp:
```diff
@@ -0,0 +1,85 @@
+
+EncoderMotor::EncoderMotor(int motor_num, Encoder& encoder, bool backward) :
+  m_motor_num(motor_num), m_encoder(encoder), m_backward(backward), 
+  m_velocityPid(-0.4,0.0,0.0), m_positionPid(-0.2,0.0,-0.2)
```
lets make these configurable?

wollip commented on src/hardware/encodermotor.cpp:
```diff
@@ -0,0 +1,85 @@
+}
+
+void EncoderMotor::hold() {
+  setPosition(m_encoder.getPosition(), 20);
```
i believe 20 is the default? lets take out the second arguement

wollip replied:
this is fine

wollip commented on src/hardware/qrd.cpp:
```diff
@@ -21,14 +21,6 @@ int8_t Qrd::getTapeError(){
   const int rc_val = analogRead(RC_QRD_SENSOR_);
   const int rr_val = analogRead(RR_QRD_SENSOR_);
 
-#if DEBUG()
```
should we keep this?

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
   LCD.begin(16,2) ;
 
-  Serial.begin(9600) ;
+  Serial.begin(115200) ;
```
debug

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
-  // while(!claw.loop()) delay(100);
+  unsigned long waitUntil = millis() + loop_delay;
+  
+  enableEncoderInterrupts();
```
move line 58 to 60 to after `Insert extra variable here`. This way delays in the intialization won't affect things

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
+    if (stopbutton()) {
+      zipline.update();
+      platform.update();
+      state = 0;  // restart routine after update
```
reset `waitTime`

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
+
+// CODE
+    if (stopbutton()){
+      gate.update();
```
reset `WaitTime`

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
+    if (stopbutton())
+    {
+      platform.update();
+      state = 0;  // restart routine after updating
```
reset `WaitTime`

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
+    {
+      driver.stop();
+      pickup.update();
+      qrd.update();
```
reset `WaitTime`

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
+    
+
+    if (stopbutton()){
+      qrd.update();
```
reset `WaitTime`

wollip commented on src/main.cpp:
```diff
@@ -40,270 +26,264 @@ void setup(){
+  //// Do not touch /////////////
+  ///////////////////////////////
+    while (millis() < waitUntil) continue;
+    waitUntil += loop_delay;
```
safer to do `waitUntil = millis() + loop_delay` just in case we have some blocking code somewhere?

wollip commented on src/main.cpp:
```diff
@@ -328,81 +308,91 @@ void loop(){
+ 
+      if (mode == 0) driver.commandDriveStraight(distance);
+      else if (mode == 1) driver.commandTurnRight(distance);
+      else driver.commandTurnLeft(distance);
```
reset `WaitTime`

wollip commented on src/main.cpp:
```diff
@@ -328,81 +308,91 @@ void loop(){
+      LCD.setCursor(7,0); LCD.print(state);
+      LCD.setCursor(0,1); LCD.print("left");
+      LCD.setCursor(7,1); LCD.print(left);
+      delay(100);
```
reset `WaitTime`

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+}
+
+bool Drivetrain::readyForCommand() {
+  return m_command == IDLE;
```
i think line follow and beacon follow counts as ready for command states

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+}
+
+void Drivetrain::tick() {
+  if (m_command == IDLE) {
```
maybe use a switch case to look nicer?

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+  else if (m_command == DRIVE_ENCODER) {
+    bool leftDone = m_leftMotor.setPosition(m_leftMotorTarget, m_speed);
+    bool rightDone = m_rightMotor.setPosition(m_rightMotorTarget, m_speed);
+    if (leftDone && rightDone) m_command = IDLE;
```
If one wheel finish a lot earlier than the other wheel, then the robot would turn. This is bad

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+  else if (m_command == DRIVE_LINE_FOLLOW) {
+    m_lineFollowPower += 5;
+    if (m_lineFollowPower > m_lineFollowMaxPower) m_lineFollowPower = m_lineFollowMaxPower;
+    int16_t error = m_lineSensor.getTapeError();
```
lets use `const` for these

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+    if (m_lineFollowPower > m_lineFollowMaxPower) m_lineFollowPower = m_lineFollowMaxPower;
+    int16_t error = m_lineSensor.getTapeError();
+    int16_t steer = error * m_lineFollowGain;
+    int16_t leftPower = m_lineFollowPower - (steer > 0 ? abs(steer) : 0);
```
you don't need `abs` on this line

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+    int16_t error = m_lineSensor.getTapeError();
+    int16_t steer = error * m_lineFollowGain;
+    int16_t leftPower = m_lineFollowPower - (steer > 0 ? abs(steer) : 0);
+    int16_t rightPower = m_lineFollowPower - (steer < 0 ? abs(steer) : 0);
```
change the `-` to a `+` and remove `abs`

Commit:
Fix zipline logic to start at agent 6 tick mark

Fix commandLineFollow call in gate

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+    m_rightMotor.setPower(m_rightPower);
+  }
+  else if (m_command == DRIVE_BEACON_FOLLOW) {
+    int32_t left10k = detect_10khz(L_BEACON_SENSOR());
```
const

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+    int32_t left10k = detect_10khz(L_BEACON_SENSOR());
+    int32_t right10k = detect_10khz(R_BEACON_SENSOR());
+    int16_t error = 0;
+    if (left10k - right10k > BEACON_UNCERTAINTY()) error = BEACON_GAIN_P();
```
can we not implement the logic here again and just use `beacon`

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+  }
+}
+
+void Drivetrain::printTo(Print& p) {
```
debug

wollip commented on src/stages/gate.cpp:
```diff
@@ -12,24 +12,24 @@ bool Gate::gate_high_ = false;
-      if (encoder_.get(hardware::R_ENCODER_) - encoder_start_ < distance_){
-        follower_.loop();
+      if (encoder_.getPosition() - encoder_start_ < distance_){
+        driver_.commandLineFollow(true);
```
change `true` to `1`

wollip commented on src/utils/pid.cpp:
```diff
@@ -0,0 +1,30 @@
+  return output;
+}
+
+void PidController::printTo(Print& p) {
```
debug

wollip commented on src/utils/pid.cpp:
```diff
@@ -0,0 +1,30 @@
+  //p.print(prevOutput);
+}
+
+}
```
comment namespace

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+  p.print(m_lineFollowPid.kI);
+}
+
+}
```
namespace

wollip commented on include/utils/pid.h:
```diff
@@ -0,0 +1,22 @@
+  float prevOutput;
+};
+
+}
```
namespace

wollip commented on include/sequences/drivetrain.h:
```diff
@@ -0,0 +1,53 @@
+
+};
+
+}
```
namespace

wollip commented on include/sequences/drivetrain.h:
```diff
@@ -0,0 +1,53 @@
+  enum Command { IDLE, DRIVE_VELOCITY, DRIVE_POWER, DRIVE_ENCODER, DRIVE_LINE_FOLLOW, DRIVE_BEACON_FOLLOW };
+  Command m_command;
+
+};
```
class

wollip commented on include/hardware/encodermotor.h:
```diff
@@ -0,0 +1,39 @@
+  
+};
+
+}
```
namespace

wollip commented on include/hardware/encodermotor.h:
```diff
@@ -0,0 +1,39 @@
+  utils::PidController m_velocityPid;
+  utils::PidController m_positionPid;
+  
+};
```
class

wollip commented on include/hardware/encoder.h:
```diff
@@ -1,43 +1,30 @@
+  static const int16_t m_numVelocitySamples = 5;
+  int16_t m_velocitySamples[m_numVelocitySamples];
+  uint8_t m_currentVelocitySample;
+};
```
class

wollip commented on include/hardware/encoder.h:
```diff
@@ -1,43 +1,30 @@
+  uint8_t m_currentVelocitySample;
+};
+
+}
```
namespace

wollip reviewed (commented):
Please change

wollip commented on src/sequences/drivetrain.cpp:
```diff
@@ -0,0 +1,128 @@
+    if (leftDone && rightDone) m_command = IDLE;
+  }
+  else if (m_command == DRIVE_LINE_FOLLOW) {
+    m_lineFollowPower += 5;
```
Is this the "Acceleration limit"??!?!?!?!?!?!?

wollip commented:
I will not agree with this change until i can do this without wheelies
```
while (true) {
  drivetrain.setPower(255,255) 
  delay()
  tick()
  drivetrain.setPower(-255,255)
  delay()
  tick()
}
```

wollip reviewed (commented):
Please fix changes

Commit:
Add debug around prints and clean up code

Changed distances to be more accurate

Safer main loop

Added motor ramping that must be explicitly disabled

Removed unused line follow PidController from Drivetrain

wollip commented on include/configs.h:
```diff
@@ -33,10 +33,9 @@
-#define GAIN_P() 12
-#define GAIN_I() 0
-#define GAIN_D() 0
+#define GAIN_P() 12.0
```
they are still being used by drivetrain.cpp

wollip commented on include/hardware/encodermotor.h:
```diff
@@ -0,0 +1,39 @@
+public:
+  EncoderMotor(int motor_num, Encoder& encoder, bool backward=false);
+  void setPower(int16_t power);
+  void setVelocity(int16_t velocity, int16_t power = 255);
```
this is fine
